python/cuda_cccl/cuda/cccl/parallel/experimental/algorithms/_three_way_partition.py

```diff
@@ -0,0 +1,258 @@
+            d_temp_storage,
+            temp_storage_bytes,
+            self.d_in_cccl,
+            self.d_first_part_out_cccl,
+            self.d_second_part_out_cccl,
+            self.d_unselected_out_cccl,
+            self.d_num_selected_out_cccl,
+            self.select_first_part_op_wrapper,
+            self.select_second_part_op_wrapper,
+            num_items,
+            stream_handle,
+        )
+        return temp_storage_bytes
+
+
+@cache_with_key(make_cache_key)
+def make_three_way_partition(
+    d_in: DeviceArrayLike | IteratorBase,
+    d_first_part_out: DeviceArrayLike | IteratorBase,
+    d_second_part_out: DeviceArrayLike | IteratorBase,
+    d_unselected_out: DeviceArrayLike | IteratorBase,
+    d_num_selected_out: DeviceArrayLike | IteratorBase,
+    select_first_part_op: Callable,
+    select_second_part_op: Callable,
+):
+    """
+    Computes a device-wide three-way partition using the specified unary ``select_first_part_op`` and ``select_second_part_op`` operators.
+
+    Example:
+        Below, ``make_three_way_partition`` is used to create a three-way partition object that can be reused.
+
+        .. literalinclude:: ../../python/cuda_cccl/tests/parallel/examples/three_way_partition/three_way_partition_object.py
+            :language: python
+            :start-after: # example-begin
+
+    Args:
+        d_in: Device array or iterator containing the input sequence of data items
+        d_first_part_out: Device array or iterator to store the first part of the output
+        d_second_part_out: Device array or iterator to store the second part of the output
+        d_unselected_out: Device array or iterator to store the unselected items
+        d_num_selected_out: Device array to store the number of items selected. The total number of items selected by ``select_first_part_op`` and ``select_second_part_op`` is stored in ``d_num_selected_out[0]`` and ``d_num_selected_out[1]``, respectively.
+        select_first_part_op: Callable representing the unary operator to select the first part
+        select_second_part_op: Callable representing the unary operator to select the second part
+
+    Returns:
+        A callable object that can be used to perform the three-way partition
+    """
+    return _ThreeWayPartition(
+        d_in,
+        d_first_part_out,
+        d_second_part_out,
+        d_unselected_out,
+        d_num_selected_out,
+        select_first_part_op,
+        select_second_part_op,
+    )
+
+
+def three_way_partition(
+    d_in: DeviceArrayLike | IteratorBase,
+    d_first_part_out: DeviceArrayLike | IteratorBase,
+    d_second_part_out: DeviceArrayLike | IteratorBase,
+    d_unselected_out: DeviceArrayLike | IteratorBase,
+    d_num_selected_out: DeviceArrayLike | IteratorBase,
+    select_first_part_op: Callable,
+    select_second_part_op: Callable,
+    num_items: int,
+    stream=None,
+):
+    """
+    Performs device-wide three-way partition.
```

Comment:
Can we extend this just a bit to include a note about what exactly a three-way partition is? I understand that this PR is already passing CI, so if you provide a blurb here I'll tack it on to another PR.

Reply:
I can just add it directly